### DISCUSSION TITLE
How to prioritize paths?

### DISCUSSION THREAD

**Discussion by mauricioszabo** [wilkerlucio/pathom3#57] [Ideas]
I'm currently having trouble with priorization. In my case, I'm trying to resolve an output that have two different paths, both with different priorizations. So, in my case, what I have is:

`a` -> `b` -> `c` (`c` have priority `0`, `b` have priority `2`)
`a` -> `c` (`c` have priority `1`)

In this case, Pathom makes the first graph of prioritizations `[0 2 0`] and the second `[0 1]`, and when it sorts it chose the first path. This is mostly likely what we don't want to happen.

But, is there any "rule" that we can chose so that prioritization works? I thought that maybe we could the sorting use the resolver output like:

```
[[:a 0] [:b 2] [:c 0]]
[[:a 0] [:c 1]]
```

And then sort accordingly (so, it'll try to get the "right" priorities by checking if that output exist on other paths, and if it does, which one have greater priority, and chose correctly). But as Pathom can have multiple outputs, this becomes not so clear in other cases, and also it's not clear what to do if _both paths_ have intermediary steps that do not appear on others.

Suggestions?

**Comment by mauricioszabo** [comment]
By the way, I forgot to tell that I could reasonably solve the prioritization problem by using this as a `chose-path` code:

```clojure
(defn prioritize [{::pcp/keys [graph] :as env} or-node node-ids]
  (let [expects (-> or-node :com.wsscode.pathom3.connect.planner/expects keys)
        nodes (for [id node-ids
                    parent (pcp/node-successors graph id)
                    :let [config (pcp/node-with-resolver-config graph env parent)
                          provides (:com.wsscode.pathom3.connect.operation/provides config)]
                    :when (and provides (every? provides expects))]
                [id config])
        [id] (apply max-key #(-> % second (::pco/priority 0)) nodes)]
    (or id (first node-ids))))
```

I'll keep testing to see if this solves my problem

**Reply by wilkerlucio** [reply]
What is the `max-key` function?

**Reply by mauricioszabo** [reply]
Clojure core function:

```
╰─>$ clj
Clojure 1.10.3
user=> `max-key
clojure.core/max-key
user=> (doc max-key)
-------------------------
clojure.core/max-key
([k x] [k x y] [k x y & more])
  Returns the x for which (k x), a number, is greatest.

  If there are multiple such xs, the last one is returned.
nil
```

**Reply by wilkerlucio** [reply]
Ah nice, wasn't aware of this one :)

**Reply by wilkerlucio** [reply]
This implementation landed in Pathom 3 `2022.02.25-alpha`!!
